---
title: "DEGs_GCBarley"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r load packages}
library(tidyverse)
library(magrittr)
library(edgeR)
```

```{r Data clean up}
# load and clean up gene counts
geneCounts <- read_delim("4_Counts/GC_GABA_ABA_geneCounts_final.txt", 
                         delim = "\t", 
                         comment = "#") %>% 
  column_to_rownames("Geneid")
colnames(geneCounts) %<>% str_remove("../3_BAM/") %>% str_remove(".STARAligned_sortedByName.bam")
# crate meta data
meta <- tibble(sample = colnames(geneCounts),
              group = str_extract(sample,".+(?=_)"), # this info will fill to 'group' as col name match 'group' pram in DGEList()
              reps = str_extract(sample,"(?<=_).+")) %>% 
  filter(sample != "Kont_S2")
geneCounts %<>% dplyr::select(-Kont_S2) # remove the sample that is very badly mapped

# construct DGElist, filtering and normalize RNA composition
countList <- DGEList(counts = geneCounts,
                     samples = meta) 
keep <- filterByExpr(countList)
# By default, the function keeps genes with about 10 read counts or more in a minimum number of samples, where the number of samples is chosen according to the minimum group sample size. The actual filtering uses CPM values rather than counts in order to avoid giving preference to samples with large library sizes. For this dataset, the median library size is about 29 million and 10/29 approx. 0.34, so the filterByExpr function keeps genes that have a CPM of 0.34 or more in at least 4 samples.
table(keep)
countList <- countList[keep,, keep.lib.sizes=FALSE] %>% 
  calcNormFactors()
# set control group as reference 
countList$samples$group <- relevel(countList$samples$group, ref = "Kont")
```

```{r mds}
# visualize sample clustering, alternatively PCA can do the job
mds <- plotMDS(countList, method="bcv", col=as.numeric(countList$samples$group))
# make it easy to change appearance 
mds %>%
  extract(c(5,6)) %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  left_join(meta) %>%
  ggplot(aes(x = x, y = y)) +
  geom_point(aes(colour  = group), 
             size = 7) +
  theme_bw() +
  theme(text = element_text(size=20)) +
  labs(x = "BCV distance 1",
       y = "BCV distance 2") -> mdsplot

# export::graph2office(x = mdsplot, type = "ppt",
#              file = "Barley_MDS.pptx", width = 10,
#              aspectr = 1.5)
```

```{r estimating dispersions}
# if with multiple factors, dispersion needs to be estimated with specified design matrix (see section 2.10.2 in edgeR user's guide updated in 21 Oct. 2019)
counts_Disp <- estimateDisp(countList)
plotBCV(counts_Disp)
```

```{r model fitting}
# contract a design matrix with baseline control
designMat <- model.matrix(~ group, data = countList$samples)

# fit to model
fit <- glmQLFit(counts_Disp)
```


```{r ABA DE genes}
# ABA Vs Control
ABA <- glmQLFTest(fit, coef = 2)
ABAgenes <- ABA$table %>% 
  rownames_to_column("GeneID") %>% 
  mutate(adj.p = p.adjust(PValue, method = "BH"))

ABAgenes %>% 
  mutate(expression = ifelse(adj.p < 0.05 & abs(logFC) >= 1, 
                             ifelse(logFC >= 1 ,'Up','Down'),
                             'Stable'),
         expression = fct_relevel(expression, "Up","Stable","Down")) %>% 
  ggplot(aes(x = logFC, 
             y = -log10(adj.p), 
             colour = expression)) +
  geom_point(alpha = 0.4, size = 3.5) +
  scale_color_manual(values=c("red", "grey","blue"))+
  xlim(c(-10, 10)) +
  geom_vline(xintercept = c(-1,1),
             lty=4,col="black",
             lwd=0.8) +
  geom_hline(yintercept = -log10(0.05),
             lty=4,col="black",
             lwd=0.8) +
  labs(x="log2(fold change)",
       y="-log10 (adj.p-value)")  +
  theme_bw() +
  theme(text = element_text(size=20),
        legend.text = element_text(size = 18),
        plot.title = element_text(hjust = 0.5), 
        legend.position = c(1, 0), 
        legend.justification = c(1, 0), 
        legend.title = element_blank()) +
  ggtitle("ABA vs. Control") -> Vocano_ABA
# plotly::ggplotly is able to turn ggplot to interactive plot

# export::graph2office(x = Vocano_ABA, type = "ppt",
#              file = "Barley_vocano.pptx", width = 10,
#              aspectr = 1.5, append = TRUE)
```

```{r GABA DE genes}
# GABA Vs Control
GABA <- glmQLFTest(fit, coef = 3)
GABAgenes <- GABA$table %>% 
  rownames_to_column("GeneID") %>% 
  mutate(adj.p = p.adjust(PValue, method = "BH"))

GABAgenes %>% 
  mutate(expression = ifelse(adj.p < 0.05 & abs(logFC) >= 1, 
                             ifelse(logFC >= 1 ,'Up','Down'),
                             'Stable'),
         expression = fct_relevel(expression, "Up","Stable","Down")) %>% 
  ggplot(aes(x = logFC, 
             y = -log10(adj.p), 
             colour = expression)) +
  geom_point(alpha = 0.4, size = 3.5) +
  scale_color_manual(values=c("red", "grey","blue"))+
  xlim(c(-11, 11)) +
  geom_vline(xintercept = c(-1,1),
             lty=4,col="black",
             lwd=0.8) +
  geom_hline(yintercept = -log10(0.05),
             lty=4,col="black",
             lwd=0.8) +
  labs(x="log2(fold change)",
       y="-log10 (adj.p-value)")  +
  theme_bw() +
  theme(text = element_text(size=20),
        legend.text = element_text(size = 18),
        plot.title = element_text(hjust = 0.5), 
        legend.position = c(1, 0), 
        legend.justification = c(1, 0), 
        legend.title = element_blank()) +
  ggtitle("GABA vs. Control") -> Vocano_GABA

# export::graph2office(x = Vocano_GABA, type = "ppt",
#              file = "Barley_vocano.pptx", width = 10,
#              aspectr = 1.5, append = TRUE)
```

```{r some prepare for GO}
DEs <- list(ABAde = ABAgenes,
     GABAde = GABAgenes) %>% 
  lapply(function(x){
    x %>% 
      filter(adj.p <= 0.05 & abs(logFC) >= 1) %>% 
      mutate(expression = ifelse(logFC >= 1, 
                                 "Up", "Down"))
  })

ABAtags <- ABAgenes %>% 
  mutate(ABA = ifelse(adj.p <= 0.05 & abs(logFC) >= 1, 
                      1,0)) %>% 
  dplyr::select(GeneID,ABA,logFC) %>% 
  data.table::setnames("logFC","ABA_logFC") 
GABAtags <- GABAgenes %>% 
  mutate(GABA = ifelse(adj.p <= 0.05 & abs(logFC) >= 1, 
                       1,0)) %>% 
  dplyr::select(GeneID,GABA,logFC) %>% 
  data.table::setnames("logFC","GABA_logFC")

# vennDiagram 
merge(ABAtags,
      GABAtags,
      by = "GeneID") %>%
  dplyr::select(-contains("logFC")) %>% 
  column_to_rownames("GeneID") %>%
  vennDiagram(circle.col=c("turquoise", "salmon"))

# common genes
patternDE <- merge(ABAtags,
      GABAtags,
      by = "GeneID") %>% 
  mutate(ABA = ifelse(ABA == 1 & ABA_logFC < 0,
                      -1, ABA),
         GABA = ifelse(GABA == 1 & GABA_logFC < 0,
                      -1, GABA)) %>% 
  column_to_rownames("GeneID") %>% 
  filter(abs(ABA) == 1 | abs(GABA) == 1)
  

# same direction
patternDE %>% 
  filter((ABA == 1 & GABA == 1) | (ABA == -1 & GABA == -1)) %>% 
  row.names() # 448

# opposite direction
patternDE %>% 
  filter((ABA == 1 & GABA == -1) | (ABA == -1 & GABA == 1)) %>% 
  row.names() # 128

# heatmap of gene Exp pattern
patternDE %>% 
  dplyr::select(-contains("logFC")) %>% 
  pheatmap(cluster_cols = FALSE,
           show_rownames = FALSE) -> testplot

# give order to heatmap with logFC
ordering <- tibble(rowNum = testplot$tree_row$order,
                   order = seq(1, length(rowNum))) %>% 
  arrange(rowNum)

patternDE %>% 
  mutate(ABA_logFC = ifelse(ABA == 0, 0, ABA_logFC),
         GABA_logFC = ifelse(GABA == 0, 0, GABA_logFC)) %>% 
  dplyr::select(-GABA, -ABA) %>% 
  mutate(order = ordering$order) %>% 
  arrange(order) %>%  
  dplyr::select(-order) %>% 
  pheatmap::pheatmap(cluster_cols = FALSE,
           cluster_rows = FALSE,
           show_rownames = FALSE) 

# save.image("~/Box/Bioinfo/BarleyGCRNASeq/DEGs_removeKont_S2.RData")
```

